library/cartridge_set_needs_restart_to_configure.py

```diff
@@ -0,0 +1,155 @@
+
+    # check if instance was not started yet
+    if not os.path.exists(console_sock):
+        return helpers.ModuleRes(facts={'needs_restart': True})
+
+    try:
+        control_console = helpers.get_control_console(console_sock)
+    except helpers.CartridgeException as e:
+        allowed_errcodes = [
+            helpers.cartridge_errcodes.SOCKET_NOT_FOUND,
+            helpers.cartridge_errcodes.FAILED_TO_CONNECT_TO_SOCKET,
+            helpers.cartridge_errcodes.INSTANCE_IS_NOT_STARTED_YET
+        ]
+        if e.code in allowed_errcodes:
+            return helpers.ModuleRes(facts={'needs_restart': True})
+
+        raise e
+
+    # check if instance config was changed (except dynamic params)
+    current_instance_conf, err = read_yaml_file_section(
+        control_console,
+        instance_info['conf_file'],
+        instance_info['instance_id']
+    )
+    if err is not None:
+        return helpers.ModuleRes(failed=True, msg="Failed to read current instance config: %s" % err)
+
+    if check_conf_updated(new_instance_conf, current_instance_conf, helpers.dynamic_box_cfg_params):
+        return helpers.ModuleRes(facts={'needs_restart': True})
+
+    if not stateboard:
+        # check if default config was changed (except dynamic params)
+        current_default_conf, err = read_yaml_file_section(
+            control_console,
+            instance_info['app_conf_file'],
+            app_name
+        )
+        if err is not None:
+            return helpers.ModuleRes(failed=True, msg="Failed to read current default config: %s" % err)
+
+        new_default_conf.update({'cluster_cookie': cluster_cookie})
+        if check_conf_updated(new_default_conf, current_default_conf, helpers.dynamic_box_cfg_params):
+            return helpers.ModuleRes(facts={'needs_restart': True})
+
+    # if box.cfg wasn't called,
+    if not helpers.box_cfg_was_called(control_console):
+        return helpers.ModuleRes(facts={'needs_restart': True})
+
+    current_cfg = helpers.get_box_cfg(control_console)
+    if current_cfg is None:
+        return helpers.ModuleRes(facts={'needs_restart': True})
+
+    for param_name in helpers.dynamic_box_cfg_params:
+        new_value = None
+        if param_name in new_instance_conf:
+            new_value = new_instance_conf[param_name]
+        elif not stateboard and param_name in new_default_conf:
+            new_value = new_default_conf[param_name]
+
+        # This code is ran after attempt to change parameter in runtime
+        # If current parameter wasn't changed to the new value,
+        # it mean that instance should be restarted to apply change
+        if new_value is not None:
+            if current_cfg.get(param_name) != new_value:
+                return helpers.ModuleRes(facts={'needs_restart': True})
+
+    return helpers.ModuleRes(changed=False)
+
+
+if __name__ == '__main__':
+    helpers.execute_module(argument_spec, needs_restart_to_configure)
```

Comment:
Could it be one module that works that has two modes - checking config changes and checking package update?

Reply:
Yes, I made it